### PR TITLE
fix: upgrade event handler and tests

### PR DESCRIPTION
**Pull request by i-chvets**
https://github.com/canonical/seldon-core-operator/issues/101

Summary of changes:
- Added upgrade event handler which enables conflict resolution when applying resources during upgrade.
- Added upgrade, remove integration tests.
- Reworked integration tests to properly support upgrade integration test.
- Added two integration tests stubs (skipped) to be included in final version.
- Requirements and tox.ini had to be changed to include update chisme.

### CODE REVIEW

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 1]
```diff
@@ -291,21 +291,62 @@ def _upload_certs_to_container(self):
...
+        try:
             self.configmap_resource_handler.apply()
-        except ApiError:
-            raise ErrorWithStatus("K8S resources creation failed", BlockedStatus)
+        except ApiError as error:
+            if self._check_and_report_k8s_conflict(error) and force_conflicts:
+                # conflict detected when applying ConfigMap resources
+                # re-apply ConfigMap resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying ConfigMap resources")
+                self.logger.warning("Applying ConfigMap resources with conflict resolution")
```
[the remark this answers is not in the document]
Comment (i-chvets) [reply src/charm.py:355]: Will word it as in `kubectl` help page to be absolutely clear:
"Apply ConfigMap with forced changes against conflicts"

File: tests/integration/test_charm.py [hunk 2]
```diff
@@ -280,3 +281,58 @@ async def test_seldon_deployment(ops_test: OpsTest):
...
+    # verify that all resources that were deployed are removed
+    lightkube_client = Client()
+
+    # verify all CRDs in namespace are removed
+    crd_list = lightkube_client.list(
+        CustomResourceDefinition,
+        labels=[("app.juju.is/created-by", "seldon-controller-manager")],
+        namespace=ops_test.model.name,
+    )
+    assert is_empty(crd_list)
```
Comment (beliaev-maksim) [inline tests/integration/test_charm.py:317]: I would recommend to rewrite this pattern. My rationale is the following:
1. we can make it more readable
2. we exclude bad side effects where user can reuse iterator, but you already poped one item (see my example)
```
c = Counter(3, 5)
next(c)
# >>> 3
list(c)
# >>> [4]
```
vs

```
c = Counter(3, 5)
list(c)
# >>> [3, 4]
```

```suggestion
    assert not list(crd_list)
```

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -280,3 +281,58 @@ async def test_seldon_deployment(ops_test: OpsTest):
...
+            # other error than Not Found
+            assert False
+
+    # verify that all related Services are removed
+    svc_list = lightkube_client.list(
+        Service,
+        labels=[("app.juju.is/created-by", "seldon-controller-manager")],
+        namespace=ops_test.model.name,
+    )
+    assert is_empty(svc_list)
```
Comment (beliaev-maksim) [inline tests/integration/test_charm.py:338]:
```suggestion
    assert not list(svc_list)
```
see another comment

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -280,3 +281,58 @@ async def test_seldon_deployment(ops_test: OpsTest):
...
+def is_empty(iterator_list):
+    """Return True if supplied iterator is empty, False otherwise."""
+    try:
+        _ = next(iterator_list)
+    except StopIteration:
+        # iterator is empty
+        return True
+    return False
+
+
```
Comment (beliaev-maksim) [inline tests/integration/test_charm.py:295]:
```suggestion
```

see another comment

**Review by beliaev-maksim — CHANGES_REQUESTED** [review]

File: tests/integration/test_upgrade.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+      charm to be tested (juju refresh of local charm requires local charms to be in the same
+      path). Note that stable/1.14 version should be deployed with "trust"
+    - Refresh the deployed stable charm to the charm to be tested.
+    - Verify that charm is active and all resources are upgraded/intact.
+    """
+
+    # build the charm
+    charm_under_test = await ops_test.build_charm(".")
+
+    # download and deploy stable version of the charm
```
Comment (DnPlas) [inline tests/integration/test_upgrade.py:57]: Why do you need to download the charm? is `ops_test.model.deploy()` not working for you?

Comment (i-chvets) [reply tests/integration/test_upgrade.py:57]: I am downloading stable charm from Charmhub. Then deploying it with `.deploy()`, then upgrading it with to locally built with `.refresh()`

Comment (DnPlas) [reply tests/integration/test_upgrade.py:57]: I understand what you are doing, but I don't understand why you download it instead of just deploying it.

Comment (i-chvets) [reply tests/integration/test_upgrade.py:57]: I am downloading stable charm (`--channel=1.14/stable`) from 1.6, so I can upgrade it to the current, most latest one we are building/testing in this test.

File: tests/integration/test_charm.py (hunk 2 above)
Comment (i-chvets) [reply tests/integration/test_charm.py]: Will do.

File: tests/integration/test_upgrade.py (hunk 5 above)
Comment (i-chvets) [reply tests/integration/test_upgrade.py:57]: Waiting on comments from Juju team, will update according to their suggested.

**Review by DnPlas — APPROVED** [review]
LGTM, thanks @i-chvets

**Pull request merged**